### PR TITLE
Query config

### PR DESCRIPTION
**Pull request by jglidden**

### CODE REVIEW

**Comment by fritzo** [comment]
Is there a good reason to scrap the common `Config` format and create a new `QueryConfig` format? All other uses of the C++ `Loom` object use the standard `Config` format.

**Comment by fritzo** [comment]
Here are some reasons agains forking the `Config` format:
- `loom.util.cat` knows how to parse `Config` and will need more code and a new filename `query_config.pb.gz` to parse `QueryConfig` files
- Custom configs require extra code in `QUERY_CONFIG` and `dump_query_config`
- Config formats will have duplicated information in creating `Loom` objects.

File: loom/runner.py [hunk 1]
```diff
@@ -354,7 +357,16 @@ def query(
...
     log_out = optional_file(log_out)
-    command = ['query', root_in, requests_in, responses_out, log_out]
+    if config_in is None:
+        config_in = loom.store.get_paths(root_in)['query']['config']
+    assert os.path.exists(config_in)
+    command = [
+        'query',
+        root_in,
+        requests_in,
+        responses_out,
```
Comment (fritzo) [inline loom/runner.py]: Sort inputs before outputs:

```
...,
config_in,
responses_out,
....
```

File: loom/store.py [hunk 2]
```diff
@@ -63,6 +63,7 @@
     },
     'query': {
         'query_log': 'query_log.pbs',
+        'config': 'config.pbs.gz',
```
Comment (fritzo) [inline loom/store.py]: Either
1. use the existing `Config` format and rename to `config.pb.gz`, or
2. keep the `QueryConfig` format, rename to `query_config.pb.gz`, and update `loom.util.cat` to recognize `query_config`

File: src/query.cc [hunk 3]
```diff
@@ -52,6 +53,7 @@ int main (int argc, char ** argv)
     const char * root_in = args.pop();
     const char * requests_in = args.pop();
     const char * responses_out = args.pop();
```
Comment (fritzo) [inline src/query.cc]: Sort inputs before outputs: `config_in, responses_out`

File: src/query.cc [hunk 4]
```diff
@@ -37,6 +37,7 @@ const char * help_message =
 "\n  ROOT_IN         root dirname of dataset in loom store"
 "\n  REQUESTS_IN     filename of requests stream (e.g. requests.pbs.gz)"
 "\n  RESPONSES_OUT   filename of responses stream (e.g. responses.pbs.gz)"
+"\n  CONFIG_IN       filename of query config (e.g. query_config.pbs.gz)"
```
Comment (fritzo) [inline src/query.cc]: Sort inputs before outputs

File: loom/test/test_query.py [hunk 5]
```diff
@@ -173,3 +175,25 @@ def test_batch_score(root, model, rows, **unused):
         ]
         scores = list(server.batch_score(rows))
         assert_equal(len(scores), len(rows))
+
+
+@for_each_dataset
+def test_seed(root, model, rows, **unused):
+    requests = get_example_requests(model, rows, 'mixed')
+    with tempdir():
+        loom.config.query_config_dump({'seed': 0}, 'conf.pbs.gz')
```
Comment (fritzo) [inline loom/test/test_query.py]: Loom assumes strictly consistent filename prefixes. Either rename `conf` to `config` (or `query_config` if you really need a new format).

Comment (fritzo) [inline loom/test/test_query.py]: Also, `.pbs` means 'protobuf stream', whereas `.pb` means 'protobuf message'.

**Comment by fritzo** [comment]
A default query config should be created at `loom.tasks.ingest` and in `loom.datasets.test`. Many of the problems in this PR should be picked up by the unit tests once the config has been generated.

**Comment by jglidden** [comment]
Could you explain that? The default query config is created at loom.tasks.ingest and in loom.datasets.test.

**Comment by fritzo** [comment]
Sorry, I see you are generating them. I don't know why tests aren't failing; it must be some weird coincidence in file formats that allows a `QueryConfig` file with seed 0 to look exactly like a zero-length stream of `Config` files.

**Comment by jglidden** [comment]
I agree though that it is better to use the existing Config message type.

**Comment by fritzo** [comment]
+1

**Pull request merged**